controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+			workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 1000*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		)),
+
+		inProgressSnapshotCheckTasks: map[string]struct{}{},
+
+		syncCallback:     syncCallback,
+		proxyConnCounter: util.NewAtomicCounter(),
+	}
+
+	m.checkScheduler.SingletonModeAll()
+
+	go m.Start()
+
+	return m, nil
+}
+
+func (m *SnapshotMonitor) Start() {
+	for i := 0; i < snapshotCheckWorkerMax; i++ {
+		go m.snapshotCheckWorker(i)
+	}
+
+	go m.processSnapshotChangeEvent()
+}
+
+func (m *SnapshotMonitor) processNextEvent() bool {
+	key, quit := m.snapshotChangeEventQueue.Get()
+	if quit {
+		return false
+	}
+	defer m.snapshotChangeEventQueue.Done(key)
+
+	event := key.(SnapshotChangeEvent)
+
+	m.snapshotCheckTaskQueue.Add(snapshotCheckTask{
+		volumeName:   event.VolumeName,
+		snapshotName: event.SnapshotName,
+		changeEvent:  true,
+	})
+
+	return true
+}
+
+func (m *SnapshotMonitor) processSnapshotChangeEvent() {
+	for m.processNextEvent() {
+	}
+}
+
+func (m *SnapshotMonitor) checkSnapshots() {
+	m.logger.WithField("monitor", monitorName).Info("Starting checking snapshots")
+	defer m.logger.WithField("monitor", monitorName).Infof("Finished checking snapshots")
+
+	engines, err := m.ds.ListEnginesByNodeRO(m.nodeName)
+	if err != nil {
+		m.logger.WithField("monitor", monitorName).WithError(err).Errorf("failed to list engines on node %v", m.nodeName)
+		return
+	}
+
+	defer func() {
+		m.Lock()
+		defer m.Unlock()
+		m.LastSnapshotPeriodicCheckedAt = metav1.Time{Time: time.Now().UTC()}
+	}()
+
+	for _, engine := range engines {
+		m.logger.WithField("monitor", monitorName).Infof("Populating engine %v snapshots", engine.Name)
+		m.populateEngineSnapshots(engine)
+	}
+}
+
+func (m *SnapshotMonitor) populateEngineSnapshots(engine *longhorn.Engine) {
+	snapshots := engine.Status.Snapshots
+	for _, snapshot := range snapshots {
+		// Skip volume-head because it is not a real snapshot.
+		// A system-generated snapshot is also ignored, because the prune operations the snapshots are out of
+		// sync during replica rebuilding. More investigation is in https://github.com/longhorn/longhorn/issues/4513
+		if snapshot.Name == etypes.VolumeHeadName || !snapshot.UserCreated {
+			continue
+		}
+
+		m.snapshotCheckTaskQueue.Add(snapshotCheckTask{
+			volumeName:   engine.Spec.VolumeName,
+			snapshotName: snapshot.Name,
+			changeEvent:  false,
+		})
+	}
+}
+
+func (m *SnapshotMonitor) processNextWorkItem(id int) bool {
+	key, quit := m.snapshotCheckTaskQueue.Get()
+	if quit {
+		return false
+	}
+	defer m.snapshotCheckTaskQueue.Done(key)
+
+	task := key.(snapshotCheckTask)
+
+	dataIntegrity, err := m.getSnapshotDataIntegrity(task.volumeName)
+	if err != nil {
+		return true
+	}
+
+	if dataIntegrity == longhorn.SnapshotDataIntegrityDisabled {
+		return true
+	}
+
+	m.logger.WithField("monitor", monitorName).Debugf("Received snapshot check task %+v", task)
```

Comment:
Will this flood the longhorn manager logs in a large cluster?

Reply:
We can remove it for avoding the potential issue, because longhorn-manager uses debug level log for now.
But I think using debug-level log in a production enviroment is weird.

Reply:
Yes. You can enable this after changing the default log level to info for the longhorn managers.